File: internal/web/images.go
package web

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"m365-native/internal/chathub"
	"net/http"
	"strings"
	"time"
)

type imageGenerationRequest struct {
	Prompt         string `json:"prompt"`
	N              int    `json:"n"`
	Size           string `json:"size"`
	ResponseFormat string `json:"response_format"`
	Model          string `json:"model"`
	User           string `json:"user"`
}

func (s *Server) imageGenerations(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "method not allowed", 405)
		return
	}
	settings := serverRuntimeSettings(s)
	bodyLimit, err := requestBodyLimit(settings.TextInputLimitUTF16)
	if err != nil {
		writeOpenAIError(w, http.StatusInternalServerError, "configuration_error", err.Error())
		return
	}
	var b imageGenerationRequest
	if err := decodeBoundedJSON(w, r, bodyLimit, &b); err != nil {
		if isRequestBodyTooLarge(err) {
			writeRequestBodyTooLarge(w, r.URL.Path, bodyLimit)
			return
		}
		http.Error(w, `{"error":{"message":"prompt is required","type":"invalid_request_error"}}`, 400)
		return
	}
	if strings.TrimSpace(b.Prompt) == "" {
		http.Error(w, `{"error":{"message":"prompt is required","type":"invalid_request_error"}}`, 400)
		return
	}
	if err := validateCallerString(b.Prompt, settings.TextInputLimitUTF16); err != nil {
		writeOpenAIErrorCode(w, http.StatusBadRequest, "invalid_request_error", "text_policy_exceeded", err.Error())
		return
	}
	if b.N <= 0 {
		b.N = 1
	}
	if b.N > 4 {
		http.Error(w, "n must be between 1 and 4", 400)
		return
	}
	if b.ResponseFormat != "" && !strings.EqualFold(b.ResponseFormat, "url") && !strings.EqualFold(b.ResponseFormat, "b64_json") {
		http.Error(w, `{"error":{"message":"response_format must be url or b64_json","type":"invalid_request_error"}}`, 400)
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), time.Duration(settings.ImageTimeoutSeconds)*time.Second)
	defer cancel()
	acc, err := s.activeAccountContext(ctx)
	if err != nil {
		http.Error(w, err.Error(), 400)
		return
	}
	if acc.OID == "" || acc.TID == "" {
		acc.OID, acc.TID = extractOIDTID(acc.AccessToken)
	}
	if acc.OID == "" || acc.TID == "" {
		http.Error(w, "account missing oid/tid", 400)
		return
	}
	size := b.Size
	if size == "" {
		size = "1024x1024"
	}
	prompt := fmt.Sprintf("Generate an image with the Flux model. Size: %s. Description: %s. Return the image URL directly.", size, b.Prompt)
	res, err := s.chat.Chat(ctx, chathub.Account{AccessToken: acc.AccessToken, OID: acc.OID, TID: acc.TID}, chathub.Request{Text: prompt, Tone: "magic"})
	if err != nil {
		if writeCanonicalTerminalError(w, err) {
			return
		}
		http.Error(w, upstreamError(err), 502)
		return
	}
	log.Printf("[image-gen] images=%d text_bytes=%d events=%d raw_bytes=%d conversation_present=%t", len(res.Images), len(res.Text), len(res.Events), len(res.RawResult), res.ConversationID != "")
	if len(res.Images) == 0 {
		// Fallback: try to find image URLs in the raw result
		if urls := extractImageURLs(res.RawResult); len(urls) > 0 {
			res.Images = urls
		}
	}
	if len(res.Images) == 0 {
		// Fallback: try to find image URLs in the response text
		if urls := extractImageURLs(res.Text); len(urls) > 0 {
			res.Images = urls
		}
	}
	if len(res.Images) == 0 {
		log.Printf("[image-gen] code=no_image_resource text_bytes=%d raw_bytes=%d events=%d", len(res.Text), len(res.RawResult), len(res.Events))
		http.Error(w, `{"error":{"message":"upstream returned no image resource","type":"upstream_error"}}`, 502)
		return
	}
	images := res.Images
	if len(images) > b.N {
		images = images[:b.N]
	}
	data := make([]map[string]string, 0, len(images))
	for _, u := range images {
		if strings.EqualFold(b.ResponseFormat, "b64_json") {
			if !strings.HasPrefix(u, "data:image/") {
				http.Error(w, `{"error":{"message":"upstream returned URL, not b64_json","type":"unsupported_response_format"}}`, 502)
				return
			}
			data = append(data, map[string]string{"b64_json": strings.SplitN(u, ",", 2)[1]})
		} else {
			data = append(data, map[string]string{"url": u})
		}
	}
	jsonOut(w, map[string]any{"created": time.Now().Unix(), "data": data, "m365": map[string]any{"conversationId": res.ConversationID, "sessionId": res.SessionID, "images": images}})
}

// extractImageURLs finds image URLs in a raw JSON string by searching for URL patterns.
func extractImageURLs(raw string) []string {
	if raw == "" {
		return nil
	}
	var out []string
	seen := map[string]bool{}
	var v any
	if err := json.Unmarshal([]byte(raw), &v); err != nil {
		return nil
	}
	var walk func(any)
	walk = func(v any) {
		switch x := v.(type) {
		case []any:
			for _, e := range x {
				walk(e)
			}
		case map[string]any:
			for k, e := range x {
				lk := strings.ToLower(k)
				if s, ok := e.(string); ok && (lk == "url" || lk == "imageurl" || lk == "thumbnailurl" || lk == "downloadurl" || lk == "src" || lk == "value" || lk == "data") {
					if strings.HasPrefix(s, "https://") && !chathub.ContainsProtectedArtifactReference(s) && !seen[s] {
						if strings.Contains(strings.ToLower(s), "image") || strings.HasSuffix(strings.ToLower(s), ".png") || strings.HasSuffix(strings.ToLower(s), ".jpg") || strings.HasSuffix(strings.ToLower(s), ".jpeg") || strings.HasSuffix(strings.ToLower(s), ".webp") || strings.HasSuffix(strings.ToLower(s), ".gif") {
							seen[s] = true
							out = append(out, s)
						}
					}
				} else {
					walk(e)
				}
			}
		}
	}
	walk(v)
	return out
}

File: src/images.rs
use std::sync::Arc;

use axum::{
    Json,
    body::to_bytes,
    extract::{Request, State},
    http::{HeaderValue, StatusCode, header},
    response::{IntoResponse, Response},
};
use serde::Deserialize;
use serde_json::{Value, json};
use time::OffsetDateTime;

use crate::{
    chathub::{Account, ChatError, ChatRequest, StreamEvent},
    error::openai_error,
    web::Gateway,
};

#[derive(Deserialize)]
struct ImageRequest {
    prompt: String,
    #[serde(default)]
    n: usize,
    #[serde(default)]
    size: String,
    #[serde(default)]
    response_format: String,
}

pub(crate) async fn generations(State(gateway): State<Arc<Gateway>>, request: Request) -> Response {
    let body = match to_bytes(request.into_body(), 16 * 1024 * 1024).await {
        Ok(body) => body,
        Err(_) => return invalid("request body is too large"),
    };
    let body = match serde_json::from_slice::<ImageRequest>(&body) {
        Ok(body) if !body.prompt.trim().is_empty() => body,
        _ => return invalid("prompt is required"),
    };
    if body.prompt.encode_utf16().count() > gateway.settings.current().text_input_limit_utf16 {
        return openai_error(
            StatusCode::BAD_REQUEST,
            "invalid_request_error",
            "text_policy_exceeded",
            "輸入文字超過目前上限",
        );
    }
    let count = body.n.max(1);
    if count > 4 {
        return invalid("n must be between 1 and 4");
    }
    if !body.response_format.is_empty()
        && !matches!(
            body.response_format.to_ascii_lowercase().as_str(),
            "url" | "b64_json"
        )
    {
        return invalid("response_format must be url or b64_json");
    }
    let Some(stored) = gateway.tokens.first() else {
        return openai_error(
            StatusCode::BAD_REQUEST,
            "account_not_found",
            "account_not_found",
            "尚未登入 Microsoft 帳號",
        );
    };
    let stored = match gateway.tokens.ensure_valid(&stored.id).await {
        Ok(account) => account,
        Err(_) => {
            return openai_error(
                StatusCode::BAD_GATEWAY,
                "token_refresh_error",
                "token_refresh_error",
                "Microsoft 帳號權杖無法使用",
            );
        }
    };
    if stored.oid.is_empty() || stored.tid.is_empty() {
        return openai_error(
            StatusCode::BAD_REQUEST,
            "account_identity_error",
            "account_identity_error",
            "Microsoft 帳號缺少必要身分資訊",
        );
    }
    let size = if body.size.is_empty() {
        "1024x1024"
    } else {
        body.size.as_str()
    };
    let chat = ChatRequest {
        text: format!(
            "Generate an image with the Flux model. Size: {size}. Description: {}. Return the image URL directly.",
            body.prompt
        ),
        tone: "magic".to_owned(),
        ..ChatRequest::default()
    };
    let account = Account {
        access_token: stored.access_token,
        graph_access_token: String::new(),
        oid: stored.oid,
        tid: stored.tid,
    };
    let mut sink = |_: StreamEvent| Ok(());
    let result = match tokio::time::timeout(
        std::time::Duration::from_secs(gateway.settings.current().image_timeout_seconds),
        gateway.chat.chat(account, chat, &mut sink),
    )
    .await
    {
        Ok(Ok(result)) => result,
        Ok(Err(error)) => return chat_error(error),
        Err(_) => {
            return openai_error(
                StatusCode::GATEWAY_TIMEOUT,
                "upstream_error",
                "upstream_timeout",
                "image generation timed out",
            );
        }
    };
    let mut images = result.images;
    if images.is_empty() {
        images = extract_image_urls(&result.raw_result);
    }
    if images.is_empty() {
        images = extract_image_urls(&result.text);
    }
    if images.is_empty() {
        return openai_error(
            StatusCode::BAD_GATEWAY,
            "upstream_error",
            "no_image_resource",
            "upstream returned no image resource",
        );
    }
    images.truncate(count);
    let data = if body.response_format.eq_ignore_ascii_case("b64_json") {
        let Some(encoded) = images
            .iter()
            .map(|image| {
                image
                    .strip_prefix("data:image/")
                    .and_then(|value| value.split_once(','))
                    .map(|(_, data)| data)
            })
            .collect::<Option<Vec<_>>>()
        else {
            return openai_error(
                StatusCode::BAD_GATEWAY,
                "unsupported_response_format",
                "unsupported_response_format",
                "upstream returned URL, not b64_json",
            );
        };
        encoded
            .into_iter()
            .map(|value| json!({"b64_json": value}))
            .collect::<Vec<_>>()
    } else {
        images
            .iter()
            .map(|value| json!({"url": value}))
            .collect::<Vec<_>>()
    };
    Json(json!({
        "created": OffsetDateTime::now_utc().unix_timestamp(),
        "data": data,
        "m365": {
            "conversationId": result.conversation_id,
            "sessionId": result.session_id,
            "images": images,
        }
    }))
    .into_response()
}

fn invalid(message: &str) -> Response {
    openai_error(
        StatusCode::BAD_REQUEST,
        "invalid_request_error",
        "invalid_request_error",
        message,
    )
}

fn chat_error(error: ChatError) -> Response {
    match error {
        ChatError::RateLimited { retry_after, .. } => {
            let mut response = openai_error(
                StatusCode::TOO_MANY_REQUESTS,
                "rate_limit_error",
                "upstream_throttle",
                "ChatHub rate limited",
            );
            if let Some(retry_after) = retry_after
                && let Ok(value) = HeaderValue::from_str(&retry_after)
            {
                response.headers_mut().insert(header::RETRY_AFTER, value);
            }
            response
        }
        error => openai_error(
            StatusCode::BAD_GATEWAY,
            "upstream_error",
            "upstream_error",
            &error.to_string(),
        ),
    }
}

fn extract_image_urls(raw: &str) -> Vec<String> {
    fn walk(value: &Value, images: &mut Vec<String>) {
        match value {
            Value::Array(values) => values.iter().for_each(|value| walk(value, images)),
            Value::Object(values) => values.iter().for_each(|(key, value)| {
                if matches!(
                    key.to_ascii_lowercase().as_str(),
                    "url" | "imageurl" | "thumbnailurl" | "downloadurl" | "src" | "value" | "data"
                ) && let Some(url) = value.as_str()
                    && url.starts_with("https://")
                    && !url.contains("/v1/artifacts/")
                    && (url.to_ascii_lowercase().contains("image")
                        || [".png", ".jpg", ".jpeg", ".webp", ".gif"]
                            .iter()
                            .any(|suffix| url.to_ascii_lowercase().ends_with(suffix)))
                    && !images.iter().any(|known| known == url)
                {
                    images.push(url.to_owned());
                } else {
                    walk(value, images);
                }
            }),
            _ => {}
        }
    }

    let mut images = Vec::new();
    if let Ok(value) = serde_json::from_str(raw) {
        walk(&value, &mut images);
    }
    images
}
